server/src/main/java/org/apache/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -485,8 +486,12 @@ public void clear() throws InterruptedException
     final List<Pair<FireHydrant, SegmentIdentifier>> indexesToPersist = new ArrayList<>();
     int numPersistedRows = 0;
     long bytesPersisted = 0L;
-    for (SegmentIdentifier identifier : sinks.keySet()) {
-      final Sink sink = sinks.get(identifier);
+    Iterator<Map.Entry<SegmentIdentifier, Sink>> iterator = sinks.entrySet().iterator();
+
+    while (iterator.hasNext()) {
```

Comment:
If I understand correctly, from the code in `abandonSegment`, it looks like the sink removal should not be happening at the same time as a persist (they're both run by the `persistExecutor`).

With this change to using an iterator, I think it would now be possible to persist a Sink that was already abandoned (i.e., `abandonSegment` could remove a Sink after that Sink is returned by iterator.next() in `persistAll`)).

So I'm wondering if it would be more correct to have this "find sinks to persist" code run in one callable together with the actual persist work in the `persistExecutor` as well. 

----

There is also a code block in `push` that has a similar pattern that runs outside of the `pushExecutor` (`abandonSegment` uses `pushBarrier()`), I wonder if there are similar problems there as well:

```
final Map<SegmentIdentifier, Sink> theSinks = new HashMap<>();
    for (final SegmentIdentifier identifier : identifiers) {
      final Sink sink = sinks.get(identifier);
      if (sink == null) {
        throw new ISE("No sink for identifier: %s", identifier);
      }
      theSinks.put(identifier, sink);
      if (sink.finishWriting()) {
        totalRows.addAndGet(-sink.getNumRows());
      }
    }
```

Reply:
> So I'm wondering if it would be more correct to have this "find sinks to persist" code run in one callable together with the actual persist work in the persistExecutor as well.

you are right, `persistExecutor` using a single thread, I think it will work.

> There is also a code block in push that has a similar pattern that runs outside of the pushExecutor (abandonSegment uses pushBarrier()), I wonder if there are similar problems there as well:

I have never observed any exceptions caused by this. And I think there is no need to worry about it because the program will wait for any outstanding pushes to finish, then abandon the segment inside the persist thread:
```java
.....
// Wait for any outstanding pushes to finish, then abandon the segment inside the persist thread.
    return Futures.transform(
        pushBarrier(),
        new Function<Object, Object>()
        {
          @Nullable
          @Override
        ......
```java

Reply:
@QiuMM 

> I have never observed any exceptions caused by this. And I think there is no need to worry about it because the program will wait for any outstanding pushes to finish, then abandon the segment inside the persist thread:

I believe you're right that `push` is okay.

In the `persistAll` case I'm guessing what's happening is something like:
1. Task does an incremental publish for some sequence
2. Publish completes, `driver.registerHandoff` adds a callback after handoff finishes that calls `appenderator.drop` -> `abandonSegment`
3. Task does another incremental publish for another sequence
4. Task calls `persistAll` which iterates through sinks, but during this iteration, the handoff callback from the previous incremental publish finishes and removes a sink

For `push`, the input Sinks are read from the Sequence->Segments map instead of looking at the entire set of Sinks, and `StreamAppenderatorDriver.publish` removes the sequence's segments after the publish, so on the next incremental publish, it doesn't have the same problem of accessing Sinks that were handled/dropped as part of the previous incremental publish

Does that sound correct given what you're seeing?

Reply:
Yes, you are right. I updated this PR to make find sinks to persist run in one callable together with the actual persist work.